### ISSUE TITLE
bug in implementation

### ISSUE THREAD

**Issue by thehesiod** [RazerM/ratelimiter#3]
I believe this implementation is broken if you enter/aenter the class twice before aexit'ing (share instance with multiple classes).  The lock should be from enter/aenter through exit/aexit.  This would also mean that you can't use the sync lock in exit with aexit

**Comment by RazerM** [comment]
If the lock was held from `__enter__` through to `__exit__`, then the task being rate limited would have to execute sequentially, which is not the goal.

The rate limited task should be able to execute up to `max_calls` within `period`, but any more than that will block (while holding the lock). Once the lock is released, the other simultaneous tasks can attempt to enter until it blocks again.

Do you have an example that doesn't do what you expected?

**Comment by thehesiod** [comment]
think of this scenario:
```python
from ratelimiter import RateLimiter
import asyncio

async def worker(limiter):
    async with limiter:
        print("Done")
        await asyncio.sleep(1000)

async def main():
    limiter = RateLimiter(1, 100)

    await asyncio.gather(*[worker(limiter) for i in range(100)])


if __name__ == '__main__':
    _loop = asyncio.get_event_loop()
    _loop.run_until_complete(main())
```
they all run immediately.

**Comment by thehesiod** [comment]
this is actually an interesting problem to solve efficiently and correctly.  I think you'll need a semaphore.

**Comment by thehesiod** [comment]
here's what I've come up with so far:
```python
class RateLimiter:
    def __init__(self, max_rate: int, period_s: float or int):
        """
        Allows `max_rate` per `period_s`.  Note this implementation doesn't groom the queue (yet)
        so its memory usage will not decrease significantly

        :param max_rate: number of hits allowed per `period_s`
        :param period_s: period in seconds
        """

        assert isinstance(max_rate, int) and max_rate > 0
        assert period_s > 0

        self._max_rate = max_rate
        self._period_s = period_s
        self._lock = asyncio.Lock()

        # We'll initially allow `max_rate` to happen in parallel, and then release
        # the semaphores as new tasks can be started
        self._sema = asyncio.Semaphore(max_rate)

        # we'll push the task end-time to this queue during `__aexit__`
        self._pri_q = asyncio.PriorityQueue(max_rate)

    async def __aenter__(self):
        await self._sema.acquire()
        # by this point we can only have up to `max_rate` in parallel

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        # register finish time before we lock
        now = time.time()

        # Lock so only one task can play with the q at a time, realisitically this will
        # only lock for a significant amount of time when the queue is full, at which point
        # we can only process one task at a time anyways
        async with self._lock:
            try:
                # Start by pushing the time this task finished to the queue
                await self._pri_q.put(now)

                if not self._pri_q.full():
                    # If it's not full, we can immediately return and allow another task to start
                    return

                # Otherwise it's full so we need to wait until we can expire a q item
                # so we get the lowest priority item (oldest entry) and wait until it's
                # expired from the current window
                window_s = now - self._pri_q.get_nowait()  # nowait since we know we have an item and are locked
                sleep_s = max(0, self._period_s - window_s)  # min clipped to zero

                await asyncio.sleep(sleep_s)

                # we can now allow a request to take its place
            finally:
                self._sema.release()
```

I haven't tested it yet so it may be totally wrong :) let me know what you think.

**Comment by RazerM** [comment]
Thanks for the example and now an attempt to fix it. I'll have more time to look at it in July. I had a first attempt at fixing it using a semaphore. I also want to write tests to really nail down the expected behaviour, since they are lacking just now.

**Comment by thehesiod** [comment]
actually last version had a crucial bug, after a code review I updated it to the current impl on gist which is actually faster.

**Comment by meltinglava** [comment]
Can confirm this. Calling an API that accsept 100 calls per second. set ratelimit to 50.
```python
import aiohttp
import asyncio
from ratelimiter import RateLimiter


class callUrls:

    async_loop = asyncio.get_event_loop()
    client = aiohttp.ClientSession(loop=async_loop)
    rate = RateLimiter(max_calls=50)

    @classmethod
    async def getJSON(cls, url):
        async with cls.rate:
            async with cls.client.get(url) as response:
                info = await response.json()
                try:
                    assert response.status == 200
                except AssertionError:
                    raise AssertionError(info)
                return info

    def open_client(cls):
        cls.client = aiohttp.ClientSession(loop=cls.async_loop)
```
getting Error:
``` 
    line 20:      raise AssertionError(info)
AssertionError: {'code': 403, 'type': 'Forbidden', 'detail': 'Account Over Queries Per Second Limit'}
```

**Comment by RazerM** [comment]
I have a proof-of-concept of a ratelimiter v2 that fixes these problems, with pluggable rate limiting algorithms (e.g. [leaky bucket][lb]).

[lb]: https://en.wikipedia.org/wiki/Leaky_bucket

It'll be a v2 because the API has breaking changes. Also, it'll optionally allow limiting concurrent calls. Sometimes we only care that `__enter__` is rate limited, but sometimes we want to wait until `__exit__` until the next concurrent call can start.

I'll try push this to a v2 branch soon.

**Comment by juanformoso** [comment]
Hi, I assume this is abandoned? Cheers
